File: year-2/programming-design/java/MotorcycleContestClientServer/MotorcycleClient/src/main/java/services/BaseServiceProxy.java
package services;

import networking.Request;
import networking.Response;

import networking.ResponseType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public abstract class BaseServiceProxy {
    private static final Logger LOGGER = LogManager.getLogger(BaseServiceProxy.class);

    protected Socket connection;
    private String host;
    private int port;

    private ObjectOutputStream output;
    private ObjectInputStream input;

    private volatile boolean finished;

    private BlockingQueue<Response> responses = new ArrayBlockingQueue<Response>(10);

    public BaseServiceProxy(String host, int port) {
        LOGGER.traceEntry("setting server host = {} port = {}", host, port);
        this.host = host;
        this.port = port;
    }

    public BaseServiceProxy() {
    }

    public void setHost(String host) {
        this.host = host;
    }

    public void setPort(int port) {
        this.port = port;
    }

    protected void getConnection() throws Exception {
        LOGGER.traceEntry("ensuring connection to server");
        try {
            if (connection != null) {
                return;
            }
            LOGGER.info("creating connection");
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (IOException e) {
            LOGGER.warn("connection failed");
            throw new Exception("server is down");
        }
    }

    protected void closeConnection() {
        finished = true;
        if (connection == null) {
            return;
        }
        Socket tempConnection = connection;
        connection = null;
        LOGGER.info("closing connection");
        try {
            input.close();
            output.close();
            tempConnection.close();
            LOGGER.info("connection closed");
        } catch (IOException e) {
            LOGGER.warn("closing connection failed");
        }
    }

    protected void sendRequest(Request request) {
        LOGGER.info("sending request to server" + request);
        try {
            output.writeObject(request);
            output.flush();
        } catch (IOException e) {
            LOGGER.warn("sending request failed");
        }
        LOGGER.traceExit("request sent");
    }

    protected Response readResponse() {
        LOGGER.traceEntry("reading response from queue");
        Response response = null;
        try {
            response = responses.take();
        } catch (InterruptedException e) {
            LOGGER.warn("reading response failed");
        }
        LOGGER.info("response read " + response);
        return response;
    }

    private void startReader() {
        LOGGER.traceEntry("starting reader thread");
        Thread workerThread = new Thread(new ReaderThread());
        workerThread.start();
    }

    protected abstract void handleRegisteredParticipant(Response response);

    private class ReaderThread implements Runnable {
        public void run() {
            LOGGER.traceEntry("running reader thread");
            while (!finished) {
                try {
                    Response response = (Response) input.readObject();
                    LOGGER.info("response received from server " + response);
                    if (response.type() == ResponseType.REGISTERED_PARTICIPANT) {
                        handleRegisteredParticipant(response);
                        continue;
                    }
                    try {
                        responses.put(response);
                    } catch (InterruptedException e) {
                        LOGGER.warn("add response to queue failed");
                    }
                } catch (IOException | ClassNotFoundException e) {
                    closeConnection();
                    break;
                } catch (Exception ex) {
                    if (!finished) {
                        LOGGER.warn(ex.getMessage());
                        closeConnection();
                        break;
                    }
                }
            }
        }
    }
}

File: year2/mpp/c#/MotorcycleClientServer/MotorcycleClient/Services/BaseServiceProxy.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using log4net;
using MotorcycleCommon.Networking;
using MotorcycleCommon.Services;

namespace motorcycle_contest.Services
{
    public abstract class BaseServiceProxy
    {
        protected static readonly ILog Logger = LogManager.GetLogger("BaseServiceProxy");

        private string _host;
        private int _port;

        // private IMotorcycleObserver client;

        private NetworkStream _stream;
        private IFormatter _formatter;
        protected TcpClient _connection;

        private Queue<Response> _responses = new Queue<Response>();
        private volatile bool _finished;
        private EventWaitHandle _waitHandle;

        public BaseServiceProxy(string host, int port)
        {
            Logger.InfoFormat("creating proxy with host {0} and port {1}", host, port);
            _host = host;
            _port = port;
        }

        protected void GetConnection()
        {
            Logger.Info("ensuring connection to server");
            if (_connection != null)
            {
                Logger.Info("proxy already connected");
                return;
            }
            try
            {
                Logger.Info("creating new connection");
                _connection = new TcpClient(_host, _port);
                _stream = _connection.GetStream();
                _formatter = new BinaryFormatter();
                _finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception exception)
            {
                Logger.Warn("creating connection failed");
                throw new ServerException(exception.Message);
            }
        }

        protected void CloseConnection()
        {
            Logger.Info("closing connection to server");
            _finished = true;
            if (_connection == null)
            {
                Logger.Info("connection already closed");
                return;
            }
            TcpClient temp = _connection;
            _connection = null;
            try
            {
                _stream.Close();
                temp.Close();
                _waitHandle.Close();
                // client = null;
                Logger.Info("connection closed");
            }
            catch (Exception exception)
            {
                Logger.Warn("closing connection failed");
                throw new ServerException(exception.Message);
            }
        }

        protected void SendRequest(Request request)
        {
            try
            {
                Logger.InfoFormat("sending request {0}", request);
                _formatter.Serialize(_stream, request);
                _stream.Flush();
            }
            catch (Exception exception)
            {
                Logger.Warn("sending request failed");
                throw new ServerException(exception.Message);
            }
        }

        protected Response ReadResponse()
        {
            Logger.Info("entering ReadResponse function");
            Response response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (_responses)
                {
                    response = _responses.Dequeue();
                }
            }
            catch (Exception exception)
            {
                Logger.Warn("reading response failed");
            }
            Logger.InfoFormat("exiting read response with {0}", response);
            return response;
        }

        private void StartReader()
        {
            Logger.Info("starting reader thread");
            Thread thread = new Thread(Run);
            thread.Start();
        }

        private void Run()
        {
            Logger.Info("running reader thread");
            while (!_finished)
            {
                try
                {
                    Response response = (Response) _formatter.Deserialize(_stream);
                    Logger.InfoFormat("response recieved from server {0}", response);
                    if (response.Type == ResponseType.REGISTERED_PARTICIPANTS)
                    {
                        HandleRegisteredParticipant(response);
                    }
                    else
                    {
                        lock (_responses)
                        {
                            _responses.Enqueue(response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (SerializationException)
                {
                    CloseConnection();
                    break;
                }
                catch (Exception exception)
                {
                    Logger.Warn("reading error " + exception);
                    CloseConnection();
                    break;
                }
            }
        }

        public abstract void HandleRegisteredParticipant(Response response);
    }
}